django/events/rest_views.py

```
@@ -0,0 +1,60 @@
+from rest_framework.generics import get_object_or_404
+from events.models import Event, Theme
+from events.serializers import EventSerializer, ThemeSerializer
+from events.permissions import EventPermission
+from friendship.models import Friend, Block
+from django.db.models import Q
+from rest_framework import viewsets, permissions
+from rest_framework.decorators import action
+
+class EventViewSet(viewsets.ModelViewSet):
+    permission_classes = [permissions.IsAuthenticated, EventPermission]
+    serializer_class = EventSerializer
+    
+    def get_queryset(self):
+        """
+        Retrieves all of friends created events, own created events, and public events
+        """
+        user = self.request.user
+        friendships = Friend.objects.filter(to_user=user).values_list('from_user', flat=True)
+        blocked = Block.objects.filter(blocked=user).values_list('blocker', flat=True)
+        blocked_query = Q(visibility=True) & ~Q(creator__pk__in=blocked)
+        return Event.objects.filter(Q(creator__pk__in=friendships) 
+                                    | Q(creator__pk=user.pk) 
+                                    | blocked_query).order_by('date_created')
+    
+    def perform_create(self, serializer):
```

I wonder if it would make more sense to do this validation in the serializer class. You could write a validate_visibility(self, value) function (see docs [here](https://www.django-rest-framework.org/api-guide/serializers/#field-level-validationhttps://www.django-rest-framework.org/api-guide/serializers/#field-level-validation) and an example of how I did this [here](https://github.com/ikchu/zoe/blob/main/django/messenger/serializers.py#L26)). 

The benefits I see to doing it this way is 1) if a non-superuser attempts to make a visible event, you can raise a ValidationError and notify the user rather than failing silently and just creating the event as a non-visible one, and 2) you only have to write this check once and it'll run for both create() and update(). Whereas, if you stick with your current implementation, you'll also need override perform_update(self, serializer) with an identical body to prevent non-superusers from simply editing the visibility.